lib/api/CForecastRunner.cc

```diff
@@ -337,7 +340,7 @@ bool CForecastRunner::pushForecastJob(const std::string& controlMessage,
 
     // 2nd loop over the detectors to clone models for forecasting
     bool persistOnDisk = false;
-    if (totalMemoryUsage >= MAX_FORECAST_MODEL_MEMORY) {
+    if (totalMemoryUsage >= forecastJob.s_MaxForecastModelMemory) {
```

Comment:
If we are letting people use 500MB of memory for forecasting then there needs to be a better link to the overall job model memory.

There is already a check on line 274 that the job memory status is `ok` for forecasting to be permitted.  That means the job is using no more than 60% of its configured model memory limit.

When we were letting people use 20MB it was possible that this could cause the job's overall model memory to exceed the job-wide limit, but this would only be by a maximum of 20MB, so it was considered a reasonable risk to take.  But with this change it would be possible that a job with a 100MB total model memory limit could be told that it was allowed to use 500MB for a forecast.  Obviously it wouldn't use that much in reality - if we assume that the forecast would use no more memory than the rest of the job then it would be 60MB max in practice in this example, but that would still be causing an overrun of the configured limit that was used to decide which nodes were big enough to accommodate the job.

So I am thinking we need an additional cap on `forecastJob.s_MaxForecastModelMemory`.  Given that the soft limit definition is 60% of permitted, maybe `forecastJob.s_MaxForecastModelMemory` should be capped at 40% of the job's overall model memory limit.  This should be done in terms of values obtained from `m_ResourceMonitor` so that future code changes don't break the link between the 60% and the 40%.

Then we need to decide how this capping will be notified.  We could make it an error so that if you ask for a forecast memory limit that's inappropriately high given the job's memory limit then the forecast isn't even attempted.  This could also be checked on the Java side in advance so that it can be reported in the REST API response.  Or alternatively we could cap the limit, attempt the forecast with the capped limit, and just report an informational message about the capping.  That would be nicer in the case where the forecast still worked in reasonable time, but we should think through how we'd troubleshoot a case where a user reported slow forecasting even when they'd asked for a high forecast memory limit.  How would the person doing support know the requested forecast limit was capped.

Reply:
There are two caps.

1. The absolute largest size of the forecast (I think 500 MB is sane as that is the limit of our disk overflow)

2. The limit of forecast allowed in memory. I think making this a maximum of `40%` of the job configured model memory limit is good (or the original 20 MB, whichever is larger).


I am not 100% sure if we should throw an error or possibly have the Java side write out an audit message. I will think a bit more about support, etc.

Reply:
I think allowing the forecast to continue is good. We should indicate though that it will spool to disk and might take longer than desired. The Java side should write an audit message indicating this.

Reply:
I am going to have the C++ side throw an error if the requested memory usage is above 40% of the configured model memory max. 

On the Java side, I will adjust the limit so that it will take `std::min((0.40 * job_max_memory), requested_forecast_limit)` and write an audit message + log which limit is used. 

The C++ side will also log a warning whenever a forecast has to be spooled to disk. This is not a "failed" state, but definitely a performance concern and I think warrants a "warning" log message.